google/cloud/bigquery/_job_helpers.py

```diff
@@ -142,12 +143,24 @@ def do_query():
                 raise create_exc
 
             try:
+                # Sometimes we get a 404 after a Conflict. In this case, we
+                # have pretty high confidence that by retrying the 404, we'll
+                # (hopefully) eventually recover the job.
+                # https://github.com/googleapis/python-bigquery/issues/2134
+                get_job_retry = retry
+                if retry is not None:
+                    get_job_retry = retry.with_predicate(
+                        lambda exc: isinstance(exc, core_exceptions.NotFound)
+                        # Reference the original retry to avoid recursion.
+                        or retry._predicate(exc)
```

Comment:
@parthea Do you have any thoughts on how stable this will be? I couldn't find any public way to pull out the previous predicate from a Retry object when looking at the https://github.com/googleapis/python-api-core/blob/main/google/api_core/retry/retry_base.py code.

Reply:
An alternative to consider: do something like I did with retry versus job_retry and introduce more parameters for job_insert_retry and job_get_retry or something like that.

Reply:
Filed https://github.com/googleapis/python-api-core/issues/796, in the meantime I'll pursue an alternative that doesn't require access to private attributes.